proposals/NNNN-DXIL-Scalarization.md

```diff
@@ -0,0 +1,655 @@
+  %.i05 = extractelement <4 x float> %0, i64 0
+  %.i17 = extractelement <4 x float> %0, i64 1
+  %.i29 = extractelement <4 x float> %0, i64 2
+  %.i311 = extractelement <4 x float> %0, i64 3
+  %.i0 = extractelement <4 x i32> %1, i64 0
+  %.i01 = bitcast i32 %.i0 to float
+  %.i1 = extractelement <4 x i32> %1, i64 1
+  %.i12 = bitcast i32 %.i1 to float
+  %.i2 = extractelement <4 x i32> %1, i64 2
+  %.i23 = bitcast i32 %.i2 to float
+  %.i3 = extractelement <4 x i32> %1, i64 3
+  %.i34 = bitcast i32 %.i3 to float
+  %.i06 = fadd float %.i05, %.i01
+  %.i18 = fadd float %.i17, %.i12
+  %.i210 = fadd float %.i29, %.i23
+  %.i312 = fadd float %.i311, %.i34
+  %.upto013 = insertelement <4 x float> poison, float %.i06, i64 0
+  %.upto114 = insertelement <4 x float> %.upto013, float %.i18, i64 1
+  %.upto215 = insertelement <4 x float> %.upto114, float %.i210, i64 2
+  %3 = insertelement <4 x float> %.upto215, float %.i312, i64 3
+  ret <4 x float> %3
+}
+
+define <4 x float> @mul_float_int(<4 x float> %0, <4 x i32> %1) {
+  %.i05 = extractelement <4 x float> %0, i64 0
+  %.i17 = extractelement <4 x float> %0, i64 1
+  %.i29 = extractelement <4 x float> %0, i64 2
+  %.i311 = extractelement <4 x float> %0, i64 3
+  %.i0 = extractelement <4 x i32> %1, i64 0
+  %.i01 = bitcast i32 %.i0 to float
+  %.i1 = extractelement <4 x i32> %1, i64 1
+  %.i12 = bitcast i32 %.i1 to float
+  %.i2 = extractelement <4 x i32> %1, i64 2
+  %.i23 = bitcast i32 %.i2 to float
+  %.i3 = extractelement <4 x i32> %1, i64 3
+  %.i34 = bitcast i32 %.i3 to float
+  %.i06 = fmul float %.i05, %.i01
+  %.i18 = fmul float %.i17, %.i12
+  %.i210 = fmul float %.i29, %.i23
+  %.i312 = fmul float %.i311, %.i34
+  %.upto013 = insertelement <4 x float> poison, float %.i06, i64 0
+  %.upto114 = insertelement <4 x float> %.upto013, float %.i18, i64 1
+  %.upto215 = insertelement <4 x float> %.upto114, float %.i210, i64 2
+  %3 = insertelement <4 x float> %.upto215, float %.i312, i64 3
+  ret <4 x float> %3
+}
+```
+
+## Priority of work
+As mentioned in the background section. DXC handles 8 instructions for scalarization. We know that `BitCastInst`, `CallInst`, `LoadInst` and `StoreInst`
+Are largely handled for us if we use the `scalarizer` pass with the `-scalarize-load-store` and `-scalarize-variable-insert-extract` flags. And we probably don't need to handle the `AddrSpaceCastInst` case. Allocas could be handled by running SROA. So thats 6/8 cases
+just by using an existing pass. 
+
+There are a few ways we could go about this:
+1. support `PassInfoMixin` in the DirectX backend. 
+   - Need to override "buildCodeGenPipeline" for the DirectX backend.
+2. Add a Legacy Pass manager abstraction to `scalarizer`
+   - Do we need permission to do this? Do we need an RFC?
+3. Fork the `scalarizer` so we can have a legacy and `PassInfoMixin` in the DirectX Backend?
+   - We could make this work for dxil ops and not just intrinsics.
```

Comment:
Couldn't we also just run the pass before the backend? Is that too early?

Reply:
It needs to run after `DXILInstructionExpansion`  which is in the backend and before `DXILLowering`. So we are currently stuck in the backend unless we move the instruction Expansion.